src/clock.rs

```diff
@@ -9,18 +9,103 @@ pub trait PrciExt {
 
 impl PrciExt for PRCI {
     fn setup(self) -> ClockSetup {
-        ClockSetup { prci: self }
+        ClockSetup {
+            prci: self,
+            coreclk: None,
+        }
     }
 }
 
 pub struct ClockSetup {
-    #[allow(dead_code)]
     prci: PRCI,
+    coreclk: Option<u32>,
 }
 
 impl ClockSetup {
+    pub fn coreclk<F: Into<Hertz>>(mut self, freq: F) -> Self {
+        self.coreclk = Some(freq.into().0);
+        self
+    }
+
     pub fn freeze(self) -> Clocks {
-        Clocks { coreclk: HFXCLK }
+        let coreclk = self.coreclk.unwrap_or(HFXCLK);
+
+        // Switch to HFXCLK
+        unsafe {
+            self.prci
+                .core_clk_sel_reg
+                .modify(|r, w| w.bits(r.bits() | 1));
+            self.prci.core_clk_sel_reg.read(); // barrier
```

Comment:
What is this fence attempting to guard against?

Reply:
I think it's to make sure the previous bus transactions is complete before starting a new one. Taken from U-boot code.